### PR TITLE
Bob review

### PR DESCRIPTION
**Pull request by bschatz-swift**
Comments on branch imgr-may-10

### CODE REVIEW

File: imgr/imgrpkg/api.go [hunk 1]
```diff
@@ -3,6 +3,9 @@
 
 // Package imgrpkg implements the server side Inode Management for ProxyFS volumes.
 // The package supports to communications protocols. The balance of the package
+
+====>>> The line "The package supports to communications protocols." needs to be reworded
```
Comment (edmc-ss) [inline imgr/imgrpkg/api.go:7]: Good catch!

File: imgr/imgrpkg/api.go [hunk 2]
```diff
@@ -71,6 +74,12 @@
 //
 // This will cause the specified <volumeName> to no longer be served. Note that
 // this does not actually affect the contents of the associated Container.
+
+Does this mean that the "root" user on the system will be able to delete an active volume?
+Is that a security concern?
+
```
Comment (edmc-ss) [inline imgr/imgrpkg/api.go:80]: Absolutely anyone with access to the embedded HTTP Server can wreak havoc... but, then, they can wreak havoc by directly addressing Swift as well of course. Need we be any more secure?

Comment (bschatz-swift) [reply imgr/imgrpkg/api.go:80]: Good point!    The root user could just do "rm -rf *" or the swift CLI

As long as we are no worse than that we should be fine.

Can you tell I have been reading too many security related PLC documents lately?

File: imgr/imgrpkg/api.go (hunk 1 above)
Comment (edmc-ss) [reply imgr/imgrpkg/api.go:7]: BTW - this line wasn't commented... so it blew up the Travis-CI run...

Comment (bschatz-swift) [reply imgr/imgrpkg/api.go:7]: Yes - I know.    I wanted it to stick out and I could not find a way in github to just comment an existing file

I plan to delete this branch as soon as you are finished looking at ti.

Comment (edmc-ss) [reply imgr/imgrpkg/api.go:7]: I reworded that part to read:

// Package imgrpkg implements the server side Inode Management for ProxyFS volumes.
// While the package provides a small set of Go-callable APIs, the bulk of its
// functionality is accessed via package retryrpc-exposed RPCs. While these RPCs
// reference active volumes known to an imgrpkg instance, a RESTful API is provided
// to specify those active volumes.
//
// Note that func's listed under type RetryRPCServerStruct are the RPCs issued by
// the client side via package retryrpc connections (and, thus, not intended to be
// called directly).

And, later, since the section was then redundant:

// The RESTful API is provided by an embedded HTTP Server
// (at URL http://<PrivateIPAddr>:<HTTPServerPort>) responsing to the following:

File: imgr/imgrpkg/api.go (hunk 2 above)
Comment (edmc-ss) [reply imgr/imgrpkg/api.go:80]: > Can you tell I have been reading too many security related PLC documents lately?

:-)

One very valid concern that remains, however, is the ease with which one could do the POST API... all too easy to do that on a Container that has good data (whether or not it's ProxyFS data) in it alas... but I assume monitord will be the only one executing this "mkfs" command... and will do so knowing that it is "ok" to do...

File: imgr/imgrpkg/api.go (hunk 1 above)
Comment (bschatz-swift) [reply imgr/imgrpkg/api.go:7]: Looks good to me

**Review by edmc-ss — COMMENTED** [review]
Ugh! GitHub sucks! Unless you changed a line, I can't open up a comment on it. Terrible UI!!!

Anyway, in this Commit#14, about line 285, I was going to change that PutInodeTableEntriesStruct to be defined as follows:

type PutInodeTableEntryStruct struct {
	InodeNumber           uint64
	InodeHeadObjectNumber uint64
	InodeHeadLength       uint64
}

type PutInodeTableEntriesRequestStruct struct {
	MountID                       string
	UpdatedInodeTableEntryArray   []PutInodeTableEntryStruct
	DereferencedObjectNumberArray []uint64
}

In other words, I added two arrays. One is the PutInodeTableEntryStruct that lists the inodeNumber->{InodeHeadObjectNumber,InodeHeadLength} tuple that needs to be inserted in the InodeTable (i.e. so that a later GetInodeTableEntry can return it). The second array is just the list of Objects that can be deleted (once the checkpoint secures the corresponding InodeTable updates of course).

There's more to add (i.e. the updates to the SuperBlock that track InodeObjectCount, InodeObjectSize, and InodeBytesReferenced)... I'll add those in a moment.

I'll then see how JSON Marshaling handles this sort of thing - hopefully well. I want to stay away from anything Golang-specific just to avoid assuming that both ends of the JSON RPC is and always will be Golang :-)

**Comment by bschatz-swift** [comment]
Yeah – github sucks in this area…. Seems like people have been complaining about it for a while

Regarding the ways to encode the information, I was weighing two responses:

  1.  What existing data structures do you have?
  2.  How expensive it is to generate the formats

For item #1 “a list of InodeNumber, InodeHeadObjectNumber, InodeHeadLength tuples”.   I think your suggestion below is sufficient.   I thought about a map where InodeNumber is the key and the value is an array of struct  { InodeHeadObjectNumber, InodeHeadLength}.    That would eliminate the duplicate InodeNumber that you would be transmitting (I assume you could have duplicates for different offsets but maybe not).  It also depends if your client data structures would support this without any overhead.

For item #2, “A list of ObjectNumbers to be deleted” – seems like “DereferencedObjectNumberArray []uint64
“ is the best solution.


Thanks,

Bob

From: edmc-ss ***@***.***>
Date: Wednesday, June 2, 2021 at 4:17 PM
To: NVIDIA/proxyfs ***@***.***>
Cc: Robert Schatz ***@***.***>, Author ***@***.***>
Subject: Re: [NVIDIA/proxyfs] Bob review (#626)

@edmc-ss commented on this pull request.

Ugh! GitHub sucks! Unless you changed a line, I can't open up a comment on it. Terrible UI!!!

Anyway, in this Commit#14, about line 285, I was going to change that PutInodeTableEntriesStruct to be defined as follows:

type PutInodeTableEntryStruct struct {
InodeNumber uint64
InodeHeadObjectNumber uint64
InodeHeadLength uint64
}

type PutInodeTableEntriesRequestStruct struct {
MountID string
UpdatedInodeTableEntryArray []PutInodeTableEntryStruct
DereferencedObjectNumberArray []uint64
}

In other words, I added two arrays. One is the PutInodeTableEntryStruct that lists the inodeNumber->{InodeHeadObjectNumber,InodeHeadLength} tuple that needs to be inserted in the InodeTable (i.e. so that a later GetInodeTableEntry can return it). The second array is just the list of Objects that can be deleted (once the checkpoint secures the corresponding InodeTable updates of course).

There's more to add (i.e. the updates to the SuperBlock that track InodeObjectCount, InodeObjectSize, and InodeBytesReferenced)... I'll add those in a moment.

I'll then see how JSON Marshaling handles this sort of thing - hopefully well. I want to stay away from anything Golang-specific just to avoid assuming that both ends of the JSON RPC is and always will be Golang :-)

**Comment by edmc-ss** [comment]
Ok.. so the following worked just fine:

`
func TestJSON(t *testing.T) {
	var (
		err                                    error
		putInodeTableEntriesRequestMarshaled   []byte
		putInodeTableEntriesRequestOriginal    *PutInodeTableEntriesRequestStruct
		putInodeTableEntriesRequestUnmarshaled *PutInodeTableEntriesRequestStruct
	)

	putInodeTableEntriesRequestOriginal = &PutInodeTableEntriesRequestStruct{
		MountID: "dummy",
		UpdatedInodeTableEntryArray: []PutInodeTableEntryStruct{
			{
				InodeNumber:           1,
				InodeHeadObjectNumber: 2,
				InodeHeadLength:       3,
			},
			{
				InodeNumber:           4,
				InodeHeadObjectNumber: 5,
				InodeHeadLength:       6,
			},
		},
		DereferencedObjectNumberArray:            []uint64{7, 8, 9},
		SuperBlockInodeObjectCountAdjustment:     11,
		SuperBlockInodeObjectSizeAdjustment:      -12,
		SuperBlockInodeBytesReferencedAdjustment: 13,
	}

	fmt.Printf("putInodeTableEntriesRequestOriginal: %#v\n", putInodeTableEntriesRequestOriginal)

	putInodeTableEntriesRequestMarshaled, err = json.Marshal(putInodeTableEntriesRequestOriginal)
	if nil != err {
		t.Fatal(err)
	}

	fmt.Printf("\nputInodeTableEntriesRequestMarshaled: %s\n", utils.ByteSliceToString(putInodeTableEntriesRequestMarshaled))

	putInodeTableEntriesRequestUnmarshaled = &PutInodeTableEntriesRequestStruct{}

	err = json.Unmarshal(putInodeTableEntriesRequestMarshaled, putInodeTableEntriesRequestUnmarshaled)
	if nil != err {
		t.Fatal(err)
	}

	fmt.Printf("\nputInodeTableEntriesRequestUnmarshaled: %#v\n", putInodeTableEntriesRequestUnmarshaled)
}
`

And here is the output verifying what happened:
`
$ go test -v --run JSON
=== RUN   TestJSON
putInodeTableEntriesRequestOriginal: &imgrpkg.PutInodeTableEntriesRequestStruct{MountID:"dummy", UpdatedInodeTableEntryArray:[]imgrpkg.PutInodeTableEntryStruct{imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x1, InodeHeadObjectNumber:0x2, InodeHeadLength:0x3}, imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x4, InodeHeadObjectNumber:0x5, InodeHeadLength:0x6}}, DereferencedObjectNumberArray:[]uint64{0x7, 0x8, 0x9}, SuperBlockInodeObjectCountAdjustment:11, SuperBlockInodeObjectSizeAdjustment:-12, SuperBlockInodeBytesReferencedAdjustment:13}

putInodeTableEntriesRequestMarshaled: {"MountID":"dummy","UpdatedInodeTableEntryArray":[{"InodeNumber":1,"InodeHeadObjectNumber":2,"InodeHeadLength":3},{"InodeNumber":4,"InodeHeadObjectNumber":5,"InodeHeadLength":6}],"DereferencedObjectNumberArray":[7,8,9],"SuperBlockInodeObjectCountAdjustment":11,"SuperBlockInodeObjectSizeAdjustment":-12,"SuperBlockInodeBytesReferencedAdjustment":13}

putInodeTableEntriesRequestUnmarshaled: &imgrpkg.PutInodeTableEntriesRequestStruct{MountID:"dummy", UpdatedInodeTableEntryArray:[]imgrpkg.PutInodeTableEntryStruct{imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x1, InodeHeadObjectNumber:0x2, InodeHeadLength:0x3}, imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x4, InodeHeadObjectNumber:0x5, InodeHeadLength:0x6}}, DereferencedObjectNumberArray:[]uint64{0x7, 0x8, 0x9}, SuperBlockInodeObjectCountAdjustment:11, SuperBlockInodeObjectSizeAdjustment:-12, SuperBlockInodeBytesReferencedAdjustment:13}
--- PASS: TestJSON (0.00s)
PASS
ok  	github.com/NVIDIA/proxyfs/imgr/imgrpkg	0.316s
`

Sheesh! GitHub absolutely sucks at allowing even the most rudimentary "rich text" formatting. Ugh!

**Comment by bschatz-swift** [comment]
Great!

From: edmc-ss ***@***.***>
Date: Wednesday, June 2, 2021 at 4:41 PM
To: NVIDIA/proxyfs ***@***.***>
Cc: Robert Schatz ***@***.***>, Author ***@***.***>
Subject: Re: [NVIDIA/proxyfs] Bob review (#626)

Ok.. so the following worked just fine:

`func TestJSON(t *testing.T) {
var (
err error
putInodeTableEntriesRequestMarshaled []byte
putInodeTableEntriesRequestOriginal *PutInodeTableEntriesRequestStruct
putInodeTableEntriesRequestUnmarshaled *PutInodeTableEntriesRequestStruct
)

putInodeTableEntriesRequestOriginal = &PutInodeTableEntriesRequestStruct{

        MountID: "dummy",

        UpdatedInodeTableEntryArray: []PutInodeTableEntryStruct{

               {

                       InodeNumber:           1,

                       InodeHeadObjectNumber: 2,

                       InodeHeadLength:       3,

               },

               {

                       InodeNumber:           4,

                       InodeHeadObjectNumber: 5,

                       InodeHeadLength:       6,

               },

        },

        DereferencedObjectNumberArray:            []uint64{7, 8, 9},

        SuperBlockInodeObjectCountAdjustment:     11,

        SuperBlockInodeObjectSizeAdjustment:      -12,

        SuperBlockInodeBytesReferencedAdjustment: 13,

}



fmt.Printf("putInodeTableEntriesRequestOriginal: %#v\n", putInodeTableEntriesRequestOriginal)



putInodeTableEntriesRequestMarshaled, err = json.Marshal(putInodeTableEntriesRequestOriginal)

if nil != err {

        t.Fatal(err)

}



fmt.Printf("\nputInodeTableEntriesRequestMarshaled: %s\n", utils.ByteSliceToString(putInodeTableEntriesRequestMarshaled))



putInodeTableEntriesRequestUnmarshaled = &PutInodeTableEntriesRequestStruct{}



err = json.Unmarshal(putInodeTableEntriesRequestMarshaled, putInodeTableEntriesRequestUnmarshaled)

if nil != err {

        t.Fatal(err)

}



fmt.Printf("\nputInodeTableEntriesRequestUnmarshaled: %#v\n", putInodeTableEntriesRequestUnmarshaled)

}
`

And here is the output verifying what happened:
`$ go test -v --run JSON
=== RUN TestJSON
putInodeTableEntriesRequestOriginal: &imgrpkg.PutInodeTableEntriesRequestStruct{MountID:"dummy", UpdatedInodeTableEntryArray:[]imgrpkg.PutInodeTableEntryStruct{imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x1, InodeHeadObjectNumber:0x2, InodeHeadLength:0x3}, imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x4, InodeHeadObjectNumber:0x5, InodeHeadLength:0x6}}, DereferencedObjectNumberArray:[]uint64{0x7, 0x8, 0x9}, SuperBlockInodeObjectCountAdjustment:11, SuperBlockInodeObjectSizeAdjustment:-12, SuperBlockInodeBytesReferencedAdjustment:13}

putInodeTableEntriesRequestMarshaled: {"MountID":"dummy","UpdatedInodeTableEntryArray":[{"InodeNumber":1,"InodeHeadObjectNumber":2,"InodeHeadLength":3},{"InodeNumber":4,"InodeHeadObjectNumber":5,"InodeHeadLength":6}],"DereferencedObjectNumberArray":[7,8,9],"SuperBlockInodeObjectCountAdjustment":11,"SuperBlockInodeObjectSizeAdjustment":-12,"SuperBlockInodeBytesReferencedAdjustment":13}

putInodeTableEntriesRequestUnmarshaled: &imgrpkg.PutInodeTableEntriesRequestStruct{MountID:"dummy", UpdatedInodeTableEntryArray:[]imgrpkg.PutInodeTableEntryStruct{imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x1, InodeHeadObjectNumber:0x2, InodeHeadLength:0x3}, imgrpkg.PutInodeTableEntryStruct{InodeNumber:0x4, InodeHeadObjectNumber:0x5, InodeHeadLength:0x6}}, DereferencedObjectNumberArray:[]uint64{0x7, 0x8, 0x9}, SuperBlockInodeObjectCountAdjustment:11, SuperBlockInodeObjectSizeAdjustment:-12, SuperBlockInodeBytesReferencedAdjustment:13}
--- PASS: TestJSON (0.00s)
PASS
ok github.com/NVIDIA/proxyfs/imgr/imgrpkg 0.316s`

**Comment by edmc-ss** [comment]
Ok... going with what is discussed... Feel free to abandon (delete) this PR & branch at your liesure.

In answer to your question about if there can be any duplicate InodeNumber's in the PutInodeTableEntriesRequestStruct, there cannot be... but, since it's JSON - and not really a Golang Map - my server side implementation will have to ensure that they are not duplicated (and, of course, ensure Exclusive Leases are held on all throughout). The same is true for the DereferencedObjectNumberArray elements - no duplicates there either...

**Comment by bschatz-swift** [comment]
This was just used for a "code review" and to get around some github settings
